feat: support amule.conf overrides via amule.override.conf (Fixes #9)

nyok1912 opened the pull request:
## What
Add a simple way to customize `amule.conf`:
- New file: `/config/amule/amule.override.conf` (like amule.conf format).
- On startup, after generating the base `/config/amule/amule.conf`, merge keys from `/config/amule/amule.override.conf`:
  - Existing keys are replaced.
  - Missing keys/sections are appended.

## Why
`amule.conf` is regenerated on each startup by design; users need a persistent, declarative place to tweak only the settings they care about. Fixes #9.

## Usage
Edit `/config/amule/amule.override.conf`. Example:
```ini
[eMule]
MaxUpload=32768
MaxConnections=900
[WebServer]
UseGzip=1

nyok1912 commented:
Thanks for the proposal 🙌. The Python approach is definitely shorter and easier to read, but in this repo the bash version brings a few practical advantages that we rely on for amule.conf:

 - Preserves key order: updates are done in-place (first/last occurrence depending on policy) without moving lines.

 - Keeps comments, blank lines, and spacing exactly as in the original file.

 - Handles duplicate keys inside a section (e.g., Nick appearing more than once): updates the occurrence that actually “wins” and can optionally deduplicate earlier ones, without reordering the section.

 - Alpine/BusyBox friendly: works with BusyBox sed/grep out of the box, no extra runtime (Python) or packages required.

 - Backups only when there are changes, with rotation.

 - Change log printed (ADDED / UPDATED / UNCHANGED / DEDUPED) to ease debugging.

 - Safe with % values (e.g., DateTimeFormat=%A, %x, %X), which otherwise trigger interpolation in configparser unless explicitly disabled.

 - Minimal diffs: we’re not rewriting the whole file, so we avoid noisy diffs and reduce merge conflict risk.

By contrast, the suggested configparser snippet (as-is):

 - Rewrites the entire file on save ⇒ loses original comments, spacing, and ordering.

 - Duplicates: with strict=True it errors; with strict=False it reads them, but on write it normalizes to a single key, potentially changing semantics if ordering mattered.

 - Interpolation pitfalls: needs RawConfigParser or interpolation=None to avoid % issues.

 - No built-in conditional backup, rotation, or change summary.

 - No fine-grained control over which occurrence (first/last) gets updated.

So yes—the bash script is longer because it explicitly guarantees those behaviors (order preservation, comments, dupes, Alpine compatibility, backups). If the goal were simplicity and we didn’t care about preserving formatting/comments, the Python approach is totally valid; but for amule.conf i think is better the in-place, low-diff, comment-preserving behavior provided by the bash implementation.

What do you think?

isc30 commented:
Hi, I have been playing around with this version of the python script:
```bash
python3 - <<'EOF'
import configparser

# Read with interpolation enabled
config = configparser.ConfigParser()
config.optionxform = str
config.read("amule.conf") 

# Read override without interpolation to avoid breaking existing references
override = configparser.ConfigParser(interpolation=None)
override.optionxform = str
override.read("amule.override.conf") 

for section in override.sections():
    if not config.has_section(section):
        config.add_section(section)
    for key, value in override.items(section):
        config.set(section, key, value)

with open("amule.conf", "w") as f:
    config.write(f)
EOF
```

From what I can see, it is perfectly working for all the use cases I could think of:

- It preserves key orders (practically, key order doesn't matter for amule)
- It removes comments (good as this is an internal generated config file)
- Properly errors on invalid config like duplicated keys (prefer early errors)
- Safe with `%` values (interpolation disabled)
- Simple to read and maintain

Please let me know if you can find and use cases that break amule with this approach.
As this is an internal file the diff doesn't matter but the maintainability of the script does matter a lot.

nyok1912 commented:
I’ve been reviewing your proposal of installing Python to run the script that applies custom overrides to amule.conf…

I actually went ahead and drafted the same script in Python, and in the end it turned out to be almost as lengthy as the Bash version.

My script doesn’t just change the values of the specified keys like the Python proposal does:

 - It first makes a backup and keeps the last X backups of the config file (especially useful for debugging if any issues arise)

 - It avoids altering the order in which entries appear in the original file

 - It preserves comments

 - It preserves the original file structure

 - It generates a report of the changes applied

These may not matter to you, but I really value having a consistent order. On top of that, after applying changes on each startup it shows a report of which keys were changed, from which value to which new one.

The Bash script could certainly be much shorter if we stripped out all these features, but I consider them very valuable, they are tested, and I think they’re a good contribution to your image.

Here’s an example of the report shown at startup when changes are applied:

```docker run --rm -it -p 3000:3000 -e PUID=1000 -e PGID=1000 emulerr...
Applied overrides from: /config/amule/amule.overrides.conf
Backup created: /config/amule/amule.conf.bak-20250906-111505 (rotation: keeping latest 1)

Summary of actions:
  - UPDATED   [eMule] Nick: 'emulerr' -> 'emulerr_ok'
  - UPDATED   [eMule] MaxConnections: '1000' -> '900'
  - UPDATED   [WebServer] UseGzip: '0' -> '1'
Creating group amule with GID 1000 ...
Creating user amule with UID 1000 ...
Directory /incoming does not exists. Creating ...
Directory /temp does not exists. Creating ...
/home/amule/.aMule/amule.conf file found. Using existing configuration.```

So, considering that to keep all these features a Python script would end up almost as long anyway—and we’d also be introducing a new dependency into the container, requiring additional packages and many more hours on my part to polish it properly—does it really make sense?

Also, I should clarify that I don’t have much experience with Python, so if we want to pursue that path it would need to be someone else taking care of the implementation.

Looking forward to hearing your thoughts—cheers!
